File: src/html-utils.ts
/**
 * Utilities for extracting clean text and structured sections from SAP Note HTML content.
 * SAP Notes use <h3 class="section" id="SectionName"> headers to delimit sections.
 */

export interface NoteSection {
  heading: string;
  content: string;
}

export interface ParsedNoteContent {
  /** Extracted sections (Symptom, Solution, etc.) */
  sections: NoteSection[];
  /** Plain-text rendering of the full note (for LLMs) */
  plainText: string;
}

/**
 * Strip HTML tags and decode entities, producing clean plain text.
 */
export function stripHtml(html: string): string {
  return html
    // Replace <br> / <br/> with newlines
    .replace(/<br\s*\/?>/gi, '\n')
    // Replace </p>, </li>, </div>, </h3> with newlines
    .replace(/<\/(p|li|div|h[1-6]|tr)>/gi, '\n')
    // Replace </ol>, </ul> with newline
    .replace(/<\/(ol|ul)>/gi, '\n')
    // Replace <li> with bullet
    .replace(/<li[^>]*>/gi, '- ')
    // Remove <img> tags but keep alt text if present
    .replace(/<img[^>]*alt="([^"]*)"[^>]*>/gi, '[$1]')
    .replace(/<img[^>]*>/gi, '')
    // Remove all remaining tags
    .replace(/<[^>]+>/g, '')
    // Decode HTML entities
    .replace(/&#x00A0;/g, ' ')
    .replace(/&#34;/g, '"')
    .replace(/&#39;/g, "'")
    .replace(/&amp;/g, '&')
    .replace(/&lt;/g, '<')
    .replace(/&gt;/g, '>')
    .replace(/&quot;/g, '"')
    .replace(/&nbsp;/g, ' ')
    .replace(/&#\d+;/g, '') // remaining numeric entities
    .replace(/&[a-z]+;/gi, '') // remaining named entities
    // Clean up whitespace
    .replace(/[ \t]+/g, ' ')
    .replace(/\n[ \t]+/g, '\n')
    .replace(/\n{3,}/g, '\n\n')
    .trim();
}

/**
 * Parse SAP Note HTML content into structured sections.
 * SAP Notes use: <h3 ... id="SectionName">SectionName</h3> followed by content.
 */
export function parseNoteSections(html: string): NoteSection[] {
  const sections: NoteSection[] = [];

  // Split on <h3> section headers
  // Pattern: <h3 ... class="section" ... id="SectionName">SectionName</h3>
  const sectionRegex = /<h3[^>]*class="section"[^>]*id="([^"]*)"[^>]*>([^<]*)<\/h3>/gi;
  const matches = [...html.matchAll(sectionRegex)];

  if (matches.length === 0) {
    // No structured sections found — return entire content as one section
    const text = stripHtml(html);
    if (text) {
      sections.push({ heading: 'Content', content: text });
    }
    return sections;
  }

  for (let i = 0; i < matches.length; i++) {
    const match = matches[i];
    const heading = match[2].trim() || match[1].trim();
    const startIndex = match.index! + match[0].length;
    const endIndex = i + 1 < matches.length ? matches[i + 1].index! : html.length;

    const sectionHtml = html.substring(startIndex, endIndex);
    const content = stripHtml(sectionHtml);

    if (content) {
      sections.push({ heading, content });
    }
  }

  return sections;
}

/**
 * Parse HTML and return both structured sections and a clean plain-text rendering.
 */
export function parseNoteContent(html: string): ParsedNoteContent {
  const sections = parseNoteSections(html);

  // Build plain text with section headers
  const plainText = sections
    .map(s => `## ${s.heading}\n${s.content}`)
    .join('\n\n');

  return { sections, plainText };
}

File: py_src/mcp_sap_notes/html_utils.py
from __future__ import annotations

import re
from html import unescape


def strip_html(html: str) -> str:
    text = html
    text = re.sub(r"<br\s*/?>", "\n", text, flags=re.IGNORECASE)
    text = re.sub(r"</(p|li|div|h[1-6]|tr|ol|ul)>", "\n", text, flags=re.IGNORECASE)
    text = re.sub(r"<li[^>]*>", "- ", text, flags=re.IGNORECASE)
    text = re.sub(r"<img[^>]*alt=\"([^\"]*)\"[^>]*>", r"[\1]", text, flags=re.IGNORECASE)
    text = re.sub(r"<img[^>]*>", "", text, flags=re.IGNORECASE)
    text = re.sub(r"<[^>]+>", "", text)
    text = unescape(text)
    text = re.sub(r"[ \t]+", " ", text)
    text = re.sub(r"\n[ \t]+", "\n", text)
    text = re.sub(r"\n{3,}", "\n\n", text)
    return text.strip()


def parse_note_sections(html: str) -> list[dict[str, str]]:
    pattern = re.compile(r"<h3[^>]*class=\"section\"[^>]*id=\"([^\"]*)\"[^>]*>([^<]*)</h3>", re.IGNORECASE)
    matches = list(pattern.finditer(html))

    if not matches:
        plain = strip_html(html)
        return [{"heading": "Content", "content": plain}] if plain else []

    sections: list[dict[str, str]] = []
    for idx, match in enumerate(matches):
        heading = (match.group(2) or match.group(1) or "Content").strip()
        start = match.end()
        end = matches[idx + 1].start() if idx + 1 < len(matches) else len(html)
        section_html = html[start:end]
        content = strip_html(section_html)
        if content:
            sections.append({"heading": heading, "content": content})

    return sections


def parse_note_content(html: str) -> dict[str, object]:
    sections = parse_note_sections(html)
    plain_text = "\n\n".join(f"## {s['heading']}\n{s['content']}" for s in sections)
    return {"sections": sections, "plainText": plain_text}
